apisix/core/config_etcd.lua

```diff
@@ -314,36 +316,44 @@ local function sync_data(self)
             res.value.id = key
             self.values[pre_index] = res
             res.clean_handlers = {}
+            log.info("update data by key: ", key)
 
         else
             self.sync_times = self.sync_times + 1
             self.values[pre_index] = false
+            self.values_hash[key] = nil
+            log.info("delete data by key: ", key)
         end
 
     elseif res.value then
         res.clean_handlers = {}
         insert_tab(self.values, res)
         self.values_hash[key] = #self.values
         res.value.id = key
+        log.info("insert data by key: ", key)
     end
 
     -- avoid space waste
     -- todo: need to cover this path, it is important.
     if self.sync_times > 100 then
-        local count = 0
-        for i = 1, #self.values do
-            local val = self.values[i]
-            self.values[i] = nil
+        local values_org = table.clone(self.values)
+        table.clear(self.values)
+
+        for i = 1, #values_org do
```

Comment:
Why do we need to delete nil elements in the array? We can make judgments in this for loop：https://github.com/apache/apisix/pull/2168/files#diff-f2239cc9d5659a13ba363ebc30959916R352

Reply:
We don't want the "value" table to grow larger and larger, which will cause memory overflow.

Therefore, we need to delete the useless items in the "value" to keep it at an acceptable size.